lan867x/src/esp_eth_phy_lan867x.c

```diff
@@ -0,0 +1,279 @@
+
+
+typedef struct {
+    phy_802_3_t phy_802_3;
+} phy_lan867x_t;
+
+/***********Custom functions implementations***********/
+esp_err_t esp_eth_phy_lan867x_read_oui(phy_802_3_t *phy_802_3, uint32_t *oui)
+{
+    esp_err_t ret = ESP_OK;
+    lan867x_phyidr1_reg_t id1;
+    lan867x_phyidr2_reg_t id2;
+    esp_eth_mediator_t *eth = phy_802_3->eth;
+
+    ESP_GOTO_ON_FALSE(oui != NULL, ESP_ERR_INVALID_ARG, err, TAG, "oui can't be null");
+
+    ESP_GOTO_ON_ERROR(eth->phy_reg_read(eth, phy_802_3->addr, ETH_PHY_IDR1_REG_ADDR, &(id1.val)), err, TAG, "read ID1 failed");
+    ESP_GOTO_ON_ERROR(eth->phy_reg_read(eth, phy_802_3->addr, ETH_PHY_IDR2_REG_ADDR, &(id2.val)), err, TAG, "read ID2 failed");
+
+    *oui = (id2.oui_bits_18_23 << 18) + ((id1.oui_bits_10_17 << 10) + (id1.oui_bits_2_9<<2));
+    return ESP_OK;
+err:
+    return ret;
+}
+
+static esp_err_t lan867x_update_link_duplex_speed(phy_lan867x_t *lan867x)
+{
+    // speed is always 10 mbps, duplex is always half and link status indication is not supported. What to do ?
+    esp_err_t ret = ESP_OK;
+    esp_eth_mediator_t *eth = lan867x->phy_802_3.eth;
+    uint32_t addr = lan867x->phy_802_3.addr;
+    bmcr_reg_t bmcr;
+    ESP_GOTO_ON_ERROR(eth->phy_reg_read(eth, addr, ETH_PHY_BMCR_REG_ADDR, &(bmcr.val)), err, TAG, "read BMCR failed");
+    // to do: figure out what to do
+    eth_speed_t speed = ETH_SPEED_10M;
+    eth_duplex_t duplex = ETH_DUPLEX_HALF;
+    eth_link_t link = ETH_LINK_UP;
+    // This will be ran exactly once, when everything is setting up
+    if (lan867x->phy_802_3.link_status != link) {
+        ESP_GOTO_ON_ERROR(eth->on_state_changed(eth, ETH_STATE_SPEED, (void *)speed), err, TAG, "change speed failed");
+        ESP_GOTO_ON_ERROR(eth->on_state_changed(eth, ETH_STATE_DUPLEX, (void *)duplex), err, TAG, "change duplex failed");
+        ESP_GOTO_ON_ERROR(eth->on_state_changed(eth, ETH_STATE_LINK, (void *)link), err, TAG, "change link failed");
+        lan867x->phy_802_3.link_status = link;
+    }
+    return ESP_OK;
+err:
+    return ret;
+}
+
+static esp_err_t lan867x_get_link(esp_eth_phy_t *phy)
+{
+    esp_err_t ret = ESP_OK;
+    phy_lan867x_t *lan867x = __containerof(esp_eth_phy_into_phy_802_3(phy), phy_lan867x_t, phy_802_3);
+
+    /* Update information about link, speed, duplex */
+    ESP_GOTO_ON_ERROR(lan867x_update_link_duplex_speed(lan867x), err, TAG, "update link duplex speed failed");
+    return ESP_OK;
+err:
+    return ret;
+}
+
+static esp_err_t lan867x_init(esp_eth_phy_t *phy)
+{
+    esp_err_t ret = ESP_OK;
+    phy_802_3_t *phy_802_3 = esp_eth_phy_into_phy_802_3(phy);
+
+    /* Basic PHY init */
+    ESP_GOTO_ON_ERROR(esp_eth_phy_802_3_basic_phy_init(phy_802_3), err, TAG, "failed to init PHY");
+    /* Check PHY ID */
+    uint32_t oui;
+    uint8_t model;
+    // to do: figure out what is wrong with this OUI layout and test it
+    ESP_GOTO_ON_ERROR(esp_eth_phy_lan867x_read_oui(phy_802_3, &oui), err, TAG, "read OUI failed");
```

Comment:
Why you didn't use `esp_eth_phy_802_3_read_oui`?

Reply:
The layout of `PHY_ID1` and `PHY_ID2` register are not standard. See section 5.1.3 (page 61) and section 5.1.4 (page 62) of the datasheet.

Reply:
esp_eth_phy_802_3_read_oui returns PHY ID number combined from ID1 and ID2.